Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm

```diff
@@ -59,11 +60,201 @@
 #import <WebCore/ResourceResponse.h>
 #import <WebCore/ScrollAnimator.h>
 #import <WebCore/SharedBuffer.h>
+#import <WebCore/WebAccessibilityObjectWrapperMac.h>
 #import <wtf/cocoa/VectorCocoa.h>
 #import <wtf/text/TextStream.h>
-
 #import "PDFKitSoftLink.h"
 
+
+@interface WKAccessibilityPDFDocumentObject: NSObject {
+    RetainPtr<PDFDocument> _pdfDocument;
+    WeakObjCPtr<NSObject> _parent;
+}
+
+@property (assign) WeakPtr<WebCore::HTMLPlugInElement, WebCore::WeakPtrImplWithEventTargetData> pluginElement;
+@property (assign) WebKit::PDFPluginBase* pdfPlugin;
+
+- (id)initWithPDFDocument:(RetainPtr<PDFDocument>)document andElement:(WebCore::HTMLPlugInElement *)element;
+- (void)setPDFPluginBase:(WebKit::PDFPluginBase *)plugin;
+- (void)setPDFDocument:(RetainPtr<PDFDocument>)document;
+- (void)setParent:(NSObject *)parent;
+- (id)accessibilityHitTest:(NSPoint)point;
+- (PDFDocument*)document;
+- (NSObject *)parent;
+- (void)gotoDestination:(PDFDestination *)destination;
+// - (PDFRect) accessibilityConvertBoundsToRootView:(PDFRect)rect withPageIndex:(NSInteger)pageIndex;
+@end
+
+@implementation WKAccessibilityPDFDocumentObject
+
+@synthesize pdfPlugin = _pdfPlugin;
+@synthesize pluginElement = _pluginElement;
+
+- (id)initWithPDFDocument:(RetainPtr<PDFDocument>)document andElement:(WebCore::HTMLPlugInElement *)element
+{
+    if (!(self = [super init]))
+        return nil;
+
+    _pdfDocument = document;
+    _pluginElement = element;
+    return self;
+}
+
+- (void)setPDFPluginBase:(WebKit::PDFPluginBase *)plugin
+{
+    _pdfPlugin = plugin;
+}
+
+- (void)setPDFDocument:(RetainPtr<PDFDocument>)document
+{
+    _pdfDocument = document;
+}
+
+- (BOOL)isAccessibilityElement
+{
+    return YES;
+}
+
+- (id)accessibilityFocusedUIElement
+{
+    if (RefPtr activeAnnotation = _pdfPlugin->activeAnnotation()) {
+        if (WebCore::AXObjectCache* existingCache = _pdfPlugin->axObjectCache()) {
+            if (RefPtr object = existingCache->getOrCreate(activeAnnotation->element()))
+            ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+                return [object->wrapper() accessibilityAttributeValue:@"_AXAssociatedPluginParent"];
+            ALLOW_DEPRECATED_DECLARATIONS_END
+        }
+    }
+    for (id page in [self accessibilityChildren]) {
+        id focusedElement = [page accessibilityFocusedUIElement];
+        if (focusedElement != nil)
+            return focusedElement;
+    }
+    return nil;
+}
+
+- (PDFDocument*)document
+{
+    return _pdfDocument.get();
+}
+
+- (NSArray *)accessibilityVisibleChildren
+{
+    NSMutableArray *visiblePageElements = [NSMutableArray new];
```

Comment:
This is a leak (we don't have ARC on). Adopt it into a RetainPtr and then .autorelease() it at the end.